c/tskit/tables.c

```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+    tsk_table_collection_t *self, tsk_id_t *nodes, size_t num_nodes)
+{
+    int ret = 0;
+    tsk_id_t k, n, m, ind, pop, mut, site;
+    tsk_id_t *node_map, *individual_map, *population_map, *site_map, *mutation_map;
+    tsk_table_collection_t tables;
+
+    node_map = calloc(self->nodes.num_rows, sizeof(tsk_id_t));
+    individual_map = calloc(self->individuals.num_rows, sizeof(tsk_id_t));
+    population_map = calloc(self->populations.num_rows, sizeof(tsk_id_t));
+    site_map = calloc(self->sites.num_rows, sizeof(tsk_id_t));
+    mutation_map = calloc(self->mutations.num_rows, sizeof(tsk_id_t));
+    if (node_map == NULL || individual_map == NULL || population_map == NULL
+        || site_map == NULL || mutation_map == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    memset(node_map, 0xff, self->nodes.num_rows * sizeof(tsk_id_t));
+    memset(individual_map, 0xff, self->individuals.num_rows * sizeof(tsk_id_t));
+    memset(population_map, 0xff, self->populations.num_rows * sizeof(tsk_id_t));
+    memset(site_map, 0xff, self->sites.num_rows * sizeof(tsk_id_t));
+    memset(mutation_map, 0xff, self->mutations.num_rows * sizeof(tsk_id_t));
+
+    ret = tsk_table_collection_copy(self, &tables, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_clear(self);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // nodes, individuals, populations
+    for (k = 0; k < (tsk_id_t) num_nodes; k++) {
+        n = nodes[k];
+        node_map[n] = k;
+        ind = tables.nodes.individual[n];
+        if (ind >= 0) {
+            if (individual_map[ind] < 0) {
+                ret = individual_table_copy_row(
+                    &tables.individuals, &self->individuals, ind);
+                if (ret < 0) {
+                    goto out;
+                }
+                individual_map[ind] = ret;
+            }
+        }
+        pop = tables.nodes.population[n];
+        if (pop >= 0) {
+            if (population_map[pop] < 0) {
+                ret = population_table_copy_row(
+                    &tables.populations, &self->populations, pop);
+                if (ret < 0) {
+                    goto out;
+                }
+                population_map[tables.nodes.population[n]] = ret;
+            }
+        }
+        ret = node_table_copy_row(
+            &tables.nodes, &self->nodes, n, population_map, individual_map);
+        if (ret < 0) {
+            goto out;
+        }
+    }
+
+    // edges
+    for (k = 0; k < (tsk_id_t) tables.edges.num_rows; k++) {
+        m = node_map[tables.edges.parent[k]];
+        n = node_map[tables.edges.child[k]];
+        if ((n >= 0) && (m >= 0)) {
+            ret = edge_table_copy_row(&tables.edges, &self->edges, k, node_map);
+            if (ret < 0) {
+                goto out;
+            }
+        }
+    }
+
+    // mutations and sites
+    mut = 0;
+    for (site = 0; site < (tsk_id_t) tables.sites.num_rows; site++) {
```

Comment:
I think it would be simpler to use the ``get_row`` functions here and skip the helpers, like
```c
for (site_id = 0; ...) {
     tsk_site_table_get_row(&tables.sites, site_id, &site);
     while ((mutation_id < (tsk_id_t) tables.mutations.num_rows)
             && (tables.mutations.site[mutation_id] == site)) {
             tsk_mutation_table_get_row(&tables.mutations, mutation_id, &mutation)
             new_node = node_map[mutation.node];
             if (new_node != TSK_NULL) {
                   if (site_map[node.site] == TSK_NULL) {
                          ret = tsk_site_table_add_row(&tables.sites, site.position, /* etc */);                
                          if (ret < 0) {
                                goto out;
                          }
                         site_map[site] = ret;
                    }
                    // etc
             }
}